src/coap_client_zephyr.c

```diff
@@ -330,10 +314,15 @@ static int golioth_coap_cb(struct golioth_req_rsp *rsp)
         golioth_sys_sem_destroy(req->request_complete_ack_sem);
     }
 
+    if (req->type == GOLIOTH_COAP_REQUEST_OBSERVE)
+    {
+        // don't free observations so we can reestablish later
+        return rsp->codes.status;
+    }
+
 free_req:
     free(req);
```

Comment:
```suggestion
    if (req->type == GOLIOTH_COAP_REQUEST_OBSERVE)
    {
        free(req);
    }
```

Gate this `free()` call instead of returning above. This is safer and less likely to be accidentally missed in future changes.